File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/module/scope/Import.scala
package org.enso.compiler.core.ir.module.scope

import org.enso.compiler.core.IR
import org.enso.compiler.core.IR.{randomId, Identifier, ToStringHelper}
import org.enso.compiler.core.ir.module.Scope
import org.enso.compiler.core.ir.{
  DiagnosticStorage,
  Expression,
  IRKind,
  IdentifiedLocation,
  MetadataStorage,
  Name
}

/** Module-level import statements. */
trait Import extends Scope {

  /** @inheritdoc */
  override def mapExpressions(fn: Expression => Expression): Import

  /** @inheritdoc */
  override def setLocation(location: Option[IdentifiedLocation]): Import

  /** @inheritdoc */
  override def duplicate(
    keepLocations: Boolean   = true,
    keepMetadata: Boolean    = true,
    keepDiagnostics: Boolean = true,
    keepIdentifiers: Boolean = false
  ): Import
}

object Import {

  /** An import statement.
    *
    * @param name        the full path representing the import
    * @param rename      the name this import is visible as
    * @param isAll       is this importing exported names
    * @param onlyNames   exported names selected from the imported module
    * @param hiddenNames exported names hidden from the imported module
    * @param location    the source location that the node corresponds to
    * @param isSynthetic is this import compiler-generated
    * @param passData    the pass metadata associated with this node
    * @param diagnostics compiler diagnostics for this node
    */
  sealed case class Module(
    name: Name.Qualified,
    rename: Option[Name.Literal],
    isAll: Boolean,
    onlyNames: Option[List[Name.Literal]],
    hiddenNames: Option[List[Name.Literal]],
    override val location: Option[IdentifiedLocation],
    isSynthetic: Boolean                        = false,
    override val passData: MetadataStorage      = MetadataStorage(),
    override val diagnostics: DiagnosticStorage = DiagnosticStorage()
  ) extends Import
      with IRKind.Primitive {
    override protected var id: Identifier = randomId

    /** Creates a copy of `this`.
      *
      * @param name        the full `.`-separated path representing the import
      * @param rename      the name this import is visible as
      * @param isAll       is this importing exported names
      * @param onlyNames   exported names selected from the imported module
      * @param hiddenNames exported names hidden from the imported module
      * @param location    the source location that the node corresponds to
      * @param isSynthetic is this import compiler-generated
      * @param passData    the pass metadata associated with this node
      * @param diagnostics compiler diagnostics for this node
      * @param id          the identifier for the new node
      * @return a copy of `this`, updated with the specified values
      */
    def copy(
      name: Name.Qualified                    = name,
      rename: Option[Name.Literal]            = rename,
      isAll: Boolean                          = isAll,
      onlyNames: Option[List[Name.Literal]]   = onlyNames,
      hiddenNames: Option[List[Name.Literal]] = hiddenNames,
      location: Option[IdentifiedLocation]    = location,
      isSynthetic: Boolean                    = isSynthetic,
      passData: MetadataStorage               = passData,
      diagnostics: DiagnosticStorage          = diagnostics,
      id: Identifier                          = id
    ): Module = {
      val res = Module(
        name,
        rename,
        isAll,
        onlyNames,
        hiddenNames,
        location,
        isSynthetic,
        passData,
        diagnostics
      )
      res.id = id
      res
    }

    /** @inheritdoc */
    override def duplicate(
      keepLocations: Boolean   = true,
      keepMetadata: Boolean    = true,
      keepDiagnostics: Boolean = true,
      keepIdentifiers: Boolean = false
    ): Module =
      copy(
        location = if (keepLocations) location else None,
        passData = if (keepMetadata) passData.duplicate else MetadataStorage(),
        diagnostics =
          if (keepDiagnostics) diagnostics.copy else DiagnosticStorage(),
        id = if (keepIdentifiers) id else randomId
      )

    /** @inheritdoc */
    override def setLocation(
      location: Option[IdentifiedLocation]
    ): Module =
      copy(location = location)

    /** @inheritdoc */
    override def mapExpressions(
      fn: Expression => Expression
    ): Module = this

    /** @inheritdoc */
    override def toString: String =
      s"""
         |Module.Scope.Import.Module(
         |name = $name,
         |rename = $rename,
         |onlyNames = $onlyNames,
         |hiddenNames = $hiddenNames,
         |isAll = $isAll,
         |location = $location,
         |passData = ${this.showPassData},
         |diagnostics = $diagnostics,
         |id = $id
         |)
         |""".toSingleLine

    /** @inheritdoc */
    override def children: List[IR] =
      name :: List(
        rename.toList,
        onlyNames.getOrElse(List()),
        hiddenNames.getOrElse(List())
      ).flatten

    /** @inheritdoc */
    override def showCode(indent: Int): String = {
      val renameCode = rename.map(n => s" as ${n.name}").getOrElse("")
      if (isAll) {
        val onlyPart = onlyNames
          .map(names => " " + names.map(_.name).mkString(", "))
          .getOrElse("")
        val hidingPart = hiddenNames
          .map(names => s" hiding ${names.map(_.name).mkString(", ")}")
          .getOrElse("")
        val all = if (onlyNames.isDefined) "" else " all"
        s"from ${name.name}$renameCode import$onlyPart$all$hidingPart"
      } else {
        s"import ${name.name}$renameCode"
      }
    }

    /** Gets the name of the module visible in this scope, either the
      * original name or the rename.
      *
      * @return the name of this import visible in code
      */
    def getSimpleName: Name = rename.getOrElse(name.parts.last)

    /** Checks whether the import statement allows use of the given
      * exported name.
      *
      * Note that it does not verify if the name is actually exported
      * by the module, only checks if it is syntactically allowed.
      *
      * @param name the name to check
      * @return whether the name could be accessed or not
      */
    def allowsAccess(name: String): Boolean = {
      if (!isAll) return false;
      if (onlyNames.isDefined) {
        onlyNames.get.exists(_.name == name)
      } else if (hiddenNames.isDefined) {
        !hiddenNames.get.exists(_.name == name)
      } else {
        true
      }
    }
  }

}

File: engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/module/scope/Import.java
package org.enso.compiler.core.ir.module.scope;

import java.util.function.Function;
import org.enso.compiler.core.ir.DiagnosticStorage;
import org.enso.compiler.core.ir.Expression;
import org.enso.compiler.core.ir.IRKind;
import org.enso.compiler.core.ir.IdentifiedLocation;
import org.enso.compiler.core.ir.MetadataStorage;
import org.enso.compiler.core.ir.Name;
import org.enso.compiler.core.ir.Name.Literal;
import org.enso.compiler.core.ir.Name.Qualified;
import org.enso.compiler.core.ir.module.Scope;
import org.enso.runtime.parser.dsl.GenerateFields;
import org.enso.runtime.parser.dsl.GenerateIR;
import org.enso.runtime.parser.dsl.IRChild;
import org.enso.runtime.parser.dsl.IRField;
import scala.Option;
import scala.collection.immutable.List;

public interface Import extends Scope {

  @Override
  Import mapExpressions(Function<Expression, Expression> fn);

  @Override
  Import setLocation(Option<IdentifiedLocation> location);

  @Override
  Import duplicate(
      boolean keepLocations,
      boolean keepMetadata,
      boolean keepDiagnostics,
      boolean keepIdentifiers);

  @GenerateIR(interfaces = {Import.class, IRKind.Primitive.class})
  final class Module extends ImportModuleGen {
    @GenerateFields
    public Module(
        @IRChild Qualified name,
        @IRChild Option<Literal> rename,
        @IRField boolean isAll,
        @IRChild Option<List<Literal>> onlyNames,
        @IRChild Option<List<Literal>> hiddenNames,
        @IRField boolean isSynthetic,
        IdentifiedLocation identifiedLocation,
        MetadataStorage passData,
        DiagnosticStorage diagnostics) {
      super(
          name,
          rename,
          isAll,
          onlyNames,
          hiddenNames,
          isSynthetic,
          identifiedLocation,
          passData,
          diagnostics);
    }

    public static Builder builder() {
      return new Builder();
    }

    public static Builder builder(Module module) {
      return new Builder(module);
    }

    public Module(
        Qualified name,
        Option<Literal> rename,
        boolean isAll,
        Option<List<Literal>> onlyNames,
        Option<List<Literal>> hiddenNames,
        boolean isSynthetic,
        IdentifiedLocation identifiedLocation,
        MetadataStorage passData) {
      this(
          name,
          rename,
          isAll,
          onlyNames,
          hiddenNames,
          isSynthetic,
          identifiedLocation,
          passData,
          null);
    }

    public static Module createSynthetic(Name.Qualified name) {
      return new Module(
          name,
          Option.empty(),
          false,
          Option.empty(),
          Option.empty(),
          true,
          null,
          new MetadataStorage(),
          null);
    }

    public Module copyWithNameAndRename(Name.Qualified name, Option<Name.Literal> rename) {
      return new Builder(this).name(name).rename(rename).build();
    }

    public Module copyWithName(Name.Qualified name) {
      return new Builder(this).name(name).build();
    }

    @Override
    public String showCode(int indent) {
      var renameCode = "";
      if (rename().isDefined()) {
        renameCode = " as " + rename().get().name();
      }
      if (isAll()) {
        var onlyPart = "";
        if (onlyNames().isDefined()) {
          onlyPart = " " + onlyNames().get().map(Literal::name).mkString(", ");
        }
        var hidingPart = "";
        if (hiddenNames().isDefined()) {
          hidingPart = " hiding " + hiddenNames().get().map(Literal::name).mkString(", ");
        }
        var all = onlyNames().isDefined() ? "" : " all";
        return "from " + name().name() + renameCode + " import" + onlyPart + all + hidingPart;
      } else {
        return "import " + name().name() + renameCode;
      }
    }

    /**
     * Gets the name of the module visible in this scope, either the original name or the rename.
     *
     * @return the name of this import visible in code
     */
    public Name getSimpleName() {
      if (rename().isDefined()) {
        return rename().get();
      } else {
        return name().parts().last();
      }
    }

    /**
     * Checks whether the import statement allows use of the given exported name.
     *
     * <p>Note that it does not verify if the name is actually exported by the module, only checks
     * if it is syntactically allowed.
     *
     * @param name the name to check
     * @return whether the name could be accessed or not
     */
    public boolean allowsAccess(String name) {
      if (!isAll()) {
        return false;
      }
      if (onlyNames().isDefined()) {
        return onlyNames().get().exists(nm -> nm.name().equals(name));
      } else if (hiddenNames().isDefined()) {
        return !hiddenNames().get().exists(nm -> nm.name().equals(name));
      } else {
        return true;
      }
    }
  }
}
